woocommerce/woocommerce
# Add To Cart Button: fix style when `core/button` block isn't visible in the page

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





### Issue

The `Add To Cart Button` block uses the same classes as the `core/button` block to ensure matching styles by default. While this approach works on the editor side, it doesn’t work on the frontend. 
In the editor, all styles of **all blocks** are loaded, but on the frontend, WordPress does an optimization, loading only the styles for blocks present on the page ([source](https://make.wordpress.org/core/2021/07/01/block-styles-loading-enhancements-in-wordpress-5-8/)). This means, that the `Add To Cart Button` block has the same style of the `core/button` block only when this latter is present on the page:

https://github.com/user-attachments/assets/283a83da-d82d-40f3-91d2-b33ba4069900


## Potential Solutions

### Disable Separate Core Block Asset Loading

Set`should_load_separate_core_block_assets` to `false` disables the WordPress optimization, causing all block styles to load on the frontend side too. This would fix the issue, but it is far to be optimal. I would not go forward with this approach.

### Load Only the Button Core Block Style

The correct method involves retrieving the block style using the `get_block_style` function ([source code](https://github.com/WordPress/wordpress-develop/blob/baf408f39e92f55acb0c12f3d052e40c78586bbf/src/wp-includes/class-wp-theme-json.php#L2821-L3024)) and loading the style with `wp_add_inline_style`. This follows WordPress core approach:

![Example](https://github.com/user-attachments/assets/cf0150fa-def1-4ba2-94d1-689a605d5c19)

https://github.com/WordPress/wordpress-develop/blob/baf408f39e92f55acb0c12f3d052e40c78586bbf/src/wp-includes/global-styles-and-settings.php#L254-L349

However, `get_block_style` is a private API and should not be used outside of WordPress Core context.
Therefore, the solution adopted in this PR is rendering the `core/button` block using the `WP_Block` class without printing the output. This allow WordPress to enqueue the style of the block in the page.

## Next Steps

While this approach works, it isn't very optimal. The real solution should follow what Jetpack does:

![image](https://github.com/user-attachments/assets/a1abb0b7-729d-41bc-85c3-0542885bbebb)

Ideally, the `Add To Cart button` block should be a wrapper block around the `core/button` block. This would allow us to change the behavior of the core block attaching our custom logic.

If you agree with this approach, I'm happy to open an issue.




Closes #52626 .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable the TT5.
2. Ensure that you have the default Product Catalog Template, header and footer templates.
3. Vist the Product Catalog template on the editor Side.
4. You see that the Add To Cart Button has the border-radius.
5. Visit the `/shop` page.
6. Ensure that the Add to Cart button has the border-radius. Ensure that the frontend side style matches the editor side style.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

Add To Cart Button: fix style when `core/button` block isn't visible in the page.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
